### PR TITLE
drivers: adc: added support adc driver for lpcexpresso55s69

### LINKED ISSUE #22703: Implement ADC driver for lpcxpresso55s69
**Is your enhancement proposal related to a problem? Please describe.**
The `lpcxpresso55s69` board currently lacks support for ADC in zephyr.

**Describe the solution you'd like**
Implement a shim driver that adapts the mcux lpc ADC driver to the zephyr ADC interface. The existing `adc_mcux_adc12.c` and `adc_mcux_adc16.c` shim drivers for kinetis and i.mx rt can be used for reference. Enable `tests/drivers/adc` for this board to ensure the ADC driver builds and functions correctly.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by tobyfirth**
Added shim driver for the LPADC for lpcexpresso55s69 board.

Fixes #22703.

Signed-off-by: Toby Firth <tobyjfirth@gmail.com>

### CODE REVIEW

File: drivers/adc/Kconfig.mcux [hunk 1]
```diff
@@ -48,3 +54,26 @@ config ADC_MCUX_ADC16_VREF_ALTERNATE
 endchoice
 
 endif # ADC_MCUX_ADC16
+
+
+if ADC_MCUX_LPADC
+
+config LPADC_DO_OFFSET_CALIBRATION
+	bool "Do offset calibration"
+	default n
```
Comment (MaureenHelm) [inline drivers/adc/Kconfig.mcux:63]: You can remove this line. Bools are implicitly default n

File: drivers/adc/Kconfig.mcux [hunk 2]
```diff
@@ -48,3 +54,26 @@ config ADC_MCUX_ADC16_VREF_ALTERNATE
...
+	int "Offset value A"
+	default 10
+	help
+	  Offset value A to use if CONFIG_LPADC_DO_OFFSET_CALIBRATION is true
+
+config LPADC_OFFSET_VALUE_B
+	int "Offset value B"
+	default 10
+	help
+	  Offset value B to use if CONFIG_LPADC_DO_OFFSET_CALIBRATION is true
```
Comment (MaureenHelm) [inline drivers/adc/Kconfig.mcux:77]: I think these might be better as dts properties

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thanks for the contribution!

Please update `west.yml` to pick up the hal_nxp change, this will help fix the buildkite failure. See [here](https://docs.zephyrproject.org/latest/guides/modules.html#changes-to-existing-modules) for more info

File: drivers/adc/Kconfig.mcux (hunk 2 above)
Comment (tobyfirth) [reply drivers/adc/Kconfig.mcux]: Changed to using dts properties.

**Review by agansari — APPROVED** [review]

File: drivers/adc/adc_mcux_lpadc.c [hunk 3]
```diff
@@ -0,0 +1,401 @@
...
+		data->resolution = kLPADC_ConversionResolutionStandard;
+		break;
+	case 16:
+		data->resolution = kLPADC_ConversionResolutionHigh;
+		break;
+	default:
+		LOG_ERR("Unsupported resolution %d", sequence->resolution);
+		return -ENOTSUP;
+	}
+#endif /* FSL_FEATURE_LPADC_HAS_CMDL_MODE */
```
Comment (anangl) [inline drivers/adc/adc_mcux_lpadc.c:106]: Does it mean that if there was no `FSL_FEATURE_LPADC_HAS_CMDL_MODE`, any requested resolution would be supported?

File: drivers/adc/adc_mcux_lpadc.c [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+	lpadc_conv_result_t conv_result;
+
+#if (defined(FSL_FEATURE_LPADC_FIFO_COUNT) \
+	&& (FSL_FEATURE_LPADC_FIFO_COUNT == 2U))
+	LPADC_GetConvResult(base, &conv_result, 0U);
+#else
+	LPADC_GetConvResult(base, &conv_result);
+#endif /* FSL_FEATURE_LPADC_FIFO_COUNT */
+	uint16_t result = data->ctx.sequence.resolution < 16 ?
+			conv_result.convValue >> 3 : conv_result.convValue;
```
Comment (anangl) [inline drivers/adc/adc_mcux_lpadc.c:237]: If I understand it correctly, the 13-bit resolution case is handled here, but what about the 12-bit one?

File: drivers/adc/adc_mcux_lpadc.c [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+			  "Invalid clock source");			\
+	ASSERT_LPADC_CLK_DIV_VALID(DT_INST_PROP(n, clk_divider),	\
+		   "Invalid clock divider");			\
+	ASSERT_WITHIN_RANGE(DT_INST_PROP(n, voltage_ref), 2, 3,	\
+		"Invalid voltage reference source");	\
+	ASSERT_LPADC_CALIBRATION_AVERAGE_VALID(		\
+		DT_INST_PROP(n, calibration_average),	\
+		"Invalid converion average number for auto-calibration time");	\
+	ASSERT_WITHIN_RANGE(DT_INST_PROP(n, power_level), 1, 4,		\
+		"Invalid voltage reference source");					\
```
Comment (anangl) [inline drivers/adc/adc_mcux_lpadc.c:365]: It looks like this message was not updated after copy-pasting.

File: drivers/adc/adc_mcux_lpadc.c (hunk 3 above)
Comment (tobyfirth) [reply drivers/adc/adc_mcux_lpadc.c:106]: I have rechecked the documentation and if FSL_FEATURE_LPADC_HAS_CMDL_MODE is not enabled then there is no resolution configuration option avaible for the ADC. Which in effect means that 12/13 bit resolution in the only option. So I need to add an #else clause to the #ifdef which checks that sequence->resolution is 12/13.

File: drivers/adc/adc_mcux_lpadc.c (hunk 4 above)
Comment (tobyfirth) [reply drivers/adc/adc_mcux_lpadc.c:237]: The 12-bit case is also handled as the msb is always 0.
Hopefully this screenshot from the documentation shows what I mean and why the 12-bit case is handled here.
Screenshot from 2020-10-02 15-28-03

File: drivers/adc/adc_mcux_lpadc.c (hunk 5 above)
Comment (tobyfirth) [reply drivers/adc/adc_mcux_lpadc.c:365]: Oops, I hadn't spotted that. I will correct it.

**Comment by tobyfirth** [comment]
I have the changes for the things that @anangl brought up. I have added some comments to explain some of the behaviour around the 16-bit and 12/13-bit support in the lpadc library.

**Review by anangl — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**